clm/trainer_utils_gpu.py

```diff
@@ -0,0 +1,302 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from statistics import mean
+
+import torch
+from omegaconf.dictconfig import DictConfig
+from tqdm import tqdm
+
+from nemo.collections.nlp.data.language_modeling.megatron.megatron_batch_samplers import (
+    MegatronPretrainingRandomBatchSampler,
+)
+from nemo.utils import logging
+from nemo_aligner.metrics import InferenceMetricsHandler
+from nemo_aligner.utils.distributed import SyncTimer
+from nemo_aligner.utils.train_utils import clip_gradients
+from nemo_aligner.utils.trainer_utils import check_progress, compute_limit_batches, compute_num_steps_per_epoch
+
+from transformers import TrainingArguments
+
+IMAGE_CAPTION_KEY = "images_and_captions"
+
+
+class Trainer:
+    """trainer that implements the supervised training loop
+        this is useful for things like SFT and reward model training
+    """
+
+    def __init__(
+        self,
+        model,
+        args: TrainingArguments,
+        tokenizer,
+        train_dataset,
+        eval_dataset,
+        optimizers,
+        compute_metrics,
+        data_collator,
+        callbacks,
+    ):
+        self.model = model
+        self.test_dataloader = None
+        self.train_dataloader, self.val_dataloader = train_dataset, eval_dataset
+        
+        self.args = args
+        self.tokenizer = tokenizer
+        self.optimizer = optimizers[0]
+        self.scheduler = optimizers[1]
+
+        self.step = 0
+        self.consumed_samples = 0
+
+        self.timer = SyncTimer(
+            reduction="mean", sync_cuda=True, buffer_size=1, reduce_op=torch.distributed.ReduceOp.MAX
+        )
+
+    def setup(self, cfg, logger, ckpt_callback, run_timer, run_init_validation=False):
+        self.cfg = cfg
+        # compute `max_steps`
+        self.num_steps_per_epoch = compute_num_steps_per_epoch(
+            self.train_dataloader.batch_sampler, self.cfg.get("limit_train_batches", 1.0)
+        )
+
+        self.limit_val_batches = compute_limit_batches(len(self.val_dataloader), self.cfg.limit_val_batches)
```

Comment:
LGTM. 

`compute_limit_batches` is used to limit eval step, though we might not need this feature.